### PR TITLE
[SPARK-48461][SQL] Replace NullPointerExceptions with error class in AssertNotNull expression

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

This PR replaces `NullPointerException`s with a new error class in the `AssertNotNull` expression.

### Why are the changes needed?

We bring the advantages from the Spark error class framework to this case, enabling better user experiences and error classification.

### Does this PR introduce _any_ user-facing change?

Yes, see above.

### How was this patch tested?

This PR includes unit test coverage.

### Was this patch authored or co-authored using generative AI tooling?

GitHub copilot

### CODE REVIEW

**Comment by dtenedor** [comment]
cc @cloud-fan

**Review by cloud-fan — APPROVED** [review]

File: common/utils/src/main/resources/error/error-conditions.json [hunk 1]
```diff
@@ -3204,6 +3204,12 @@
     ],
     "sqlState" : "42809"
   },
+  "NOT_NULL_ASSERT_VIOLATION" : {
+    "message" : [
+      "NULL value appeared in non-nullable field: <walkedTypePath>If the schema is inferred from a Scala tuple/case class, or a Java bean, please try to use scala.Option[_] or other nullable types (such as java.lang.Integer instead of int/scala.Int)."
```
Comment (sigmod) [inline common/utils/src/main/resources/error/error-conditions.json:3209]: Nit?

...Path>. If

**Review by sigmod — APPROVED** [review]

File: common/utils/src/main/resources/error/error-conditions.json (hunk 1 above)
Comment (dtenedor) [reply common/utils/src/main/resources/error/error-conditions.json:3209]: The `<walkedTypePath>` contains newline(s) so we are OK here. I just ported it from the previous error message :)

Comment (sigmod) [reply common/utils/src/main/resources/error/error-conditions.json:3209]: gotcha!

**Review by HyukjinKwon — APPROVED** [review]

**Comment by dtenedor** [comment]
Alright @HyukjinKwon @cloud-fan I finally got all the tests updated :)

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**